### PR TITLE
Fixed #24941 -- Added ModelAdmin.get_exclude() hook

### PR DESCRIPTION
**Pull request by olasitarska**

### CODE REVIEW

**Comment by timgraham** [comment]
I guess this might need some additions in `django/contrib/admin/checks.py` (or maybe not, I guess the other methods don't have checks because it's impossible to check them since they aren't static).

By the way, did you see the [discussion on django-developers](https://groups.google.com/d/topic/django-developers/WrnhmTyLHuY/discussion). Maybe you could chime in there since there were some concerns about whether or not this is actually a good idea.

**Comment by olasitarska** [comment]
Hi @timgraham! 

Looking at `checks.py`, I think it doesn't check any of the dynamic methods, because as far as I understand, checks are run only once to verify that the configuration of admin in your code is ok. `get_exclude` or `get_fields`, etc, are used dynamically, so there is no way to verify that through checks. 

I'm gonna post something to django-developers, thanks. I saw the discussion there, and I think it's a worthwhile change. I think improving admin's extensibility in some small ways like that may make developers life a lot easier, so I don't see why we wouldn't introduce that change.

**Comment by timgraham** [comment]
I don't know if the `get_exclude()` method might be expensive, but it seems we could simplify by storing the result in a local variable to avoid calling it so many times.

File: docs/ref/contrib/admin/index.txt [hunk 1]
```diff
@@ -1414,6 +1414,14 @@ templates used by the :class:`ModelAdmin` views:
     being edited (or ``None`` on an add form) and is expected to return a list
     of fields, as described above in the :attr:`ModelAdmin.fields` section.
 
+.. method:: ModelAdmin.get_exclude(request, obj=None)
+
+    .. versionadded:: 1.9
+
+    The ``get_exclude`` method is given the ``HttpRequest`` and the ``obj``
+    being edited and is expected to return the attribute exclude, as described
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: expected to return a list of fields, as described ...

File: tests/modeladmin/tests.py [hunk 2]
```diff
@@ -298,6 +298,21 @@ class BandAdmin(ModelAdmin):
             list(list(ma.get_formsets_with_inlines(request))[0][0]().forms[0].fields),
             ['main_band', 'opening_band', 'day', 'id', 'DELETE'])
 
+    def test_overriding_get_exclude(self):
+        """
+        Ensure that overriding ModelAdmin.get_exclude method works
```
Comment (timgraham) [inline tests/modeladmin/tests.py]: ModelAdmin.get_exclude()
add period

File: tests/modeladmin/tests.py [hunk 3]
```diff
@@ -298,6 +298,21 @@ class BandAdmin(ModelAdmin):
...
+        Refs #24941.
```
Comment (timgraham) [inline tests/modeladmin/tests.py]: I think the ticket reference isn't needed -- the functionality is straightforward and the ticket doesn't really add additional information (we usually reserve ticket references for more obscure issues).

File: tests/modeladmin/tests.py [hunk 4]
```diff
@@ -298,6 +298,21 @@ class BandAdmin(ModelAdmin):
...
+        """
+        class BandAdmin(ModelAdmin):
+            exclude = ['bio']
+
+            def get_exclude(self, *args, **kwargs):
+                return ['name']
+
+        ma = BandAdmin(Band, self.site)
+        self.assertEqual(list(ma.get_form(request).base_fields),
```
Comment (timgraham) [inline tests/modeladmin/tests.py]: single line is fine here

File: tests/modeladmin/tests.py [hunk 5]
```diff
@@ -298,6 +298,21 @@ class BandAdmin(ModelAdmin):
...
 
+    def test_overriding_get_exclude(self):
+        """
+        Ensure that overriding ModelAdmin.get_exclude method works
+        Refs #24941.
+        """
+        class BandAdmin(ModelAdmin):
+            exclude = ['bio']
+
+            def get_exclude(self, *args, **kwargs):
```
Comment (timgraham) [inline tests/modeladmin/tests.py]: I'd use the signature from the documentation. Also, a more complex example that uses `obj` might be nice. This simple test doesn't give me great confidence that there aren't any edge cases that might cause problems. Are the changes in `get_formset()` covered by this test?

**Comment by timgraham** [comment]
A mention in the release notes is also needed.

**Comment by olasitarska** [comment]
Thanks @timgraham, I'm gonna work on this later today.

File: docs/ref/contrib/admin/index.txt [hunk 6]
```diff
@@ -1414,6 +1414,14 @@ templates used by the :class:`ModelAdmin` views:
     being edited (or ``None`` on an add form) and is expected to return a list
     of fields, as described above in the :attr:`ModelAdmin.fields` section.
 
+.. method:: ModelAdmin.get_exclude(request, obj=None)
+
+    .. versionadded:: 1.9
```
Comment (auvipy) [inline docs/ref/contrib/admin/index.txt]: this should be 1.10 now

**Comment by timgraham** [comment]
Hi Ola, any chance you could address the comments soon? If not, we can close this and ask someone else to finish it up. Thanks!

**Comment by olasitarska** [comment]
Will take a look this week Tim, sorry! Completely slipped from my radar :)

**Comment by timgraham** [comment]
Closing due to inactivity. Feel free to send a PR if you have time to return to it.

**Comment by zachborboa** [comment]
Continued this PR at https://github.com/django/django/pull/7252.

**Pull request closed without merging**